### PR TITLE
Set module output as graph output to prevent buffer reuse

### PR DESCRIPTION
**Pull request by SherlockNoMad**
Module forward output buffers were reused in the backward subgraph, the tensor value is corrupted after the execution of loss.backward(). 

image

Setting module output as graph output is one way to prevent buffer reuse, however, I don't think this is the right way... 
The buffer's ownership should be transferred to pytorch.

### CODE REVIEW

**Comment by codemzs** [comment]
@SherlockNoMad I have fixed this in a more graceful manner in my change [here](https://github.com/microsoft/onnxruntime/pull/7077/commits/e332ffe911ca7ab9c1e2c319249d573a28c6f6c5). As discussed offline I had indicated we will need to turn off reuse, it won't affect performance, this is simply an optimization to reduce calls in memory management when tensors are allocated, mostly for memory pattern generation which we are not using anyways.

**Comment by iK1D** [comment]
Can we just add another attribute something like InputsForExternal for YieldOp inputs, add one more graph travel to collect all these NodeArgs at the beginning, and treat them as graph output and use AllocKind::kAllocateOutput for them? With this way, we don't need the change the graph.

File: orttraining/orttraining/python/training/ortmodule.py [hunk 1]
```diff
@@ -236,6 +236,7 @@ def backward(ctx, *grad_outputs):
 
                     # Return input and initializer gradients
                     num_user_input_grads = len(self._input_names_require_grad)
+                    num_weigthts_to_train = len(self._onnx_graphs_info.initializer_grad_names_to_train)
```
Comment (skottmckay) [inline orttraining/orttraining/python/training/ortmodule.py:239]: nit: num_weights_to_train?

**Comment by skottmckay** [comment]
> Setting module output as graph output is one way to prevent buffer reuse, however, I don't think this is the right way...
> The buffer's ownership should be transferred to pytorch.

Not quite following this. Ownership of a graph output is passed to the caller of InferenceSession::Run. If you tag these as graph outputs what prevents ownership transferring to pytorch?
